File: ESTRUCTURA DE DATOS/DEBERES/SEGUNDO_PARCIAL/DEBER2/PROYECTO_NRC1978/LISTA_DOBLE/ListaPropietarios.cpp
#include "ListaPropietarios.h"
#include <fstream>
#include <sstream>
#include "Lista.h"

ListaPropietarios::ListaPropietarios() {
    primero = nullptr;
    ultimo = nullptr;
}

ListaPropietarios::~ListaPropietarios() {
    Nodo<Propietario>* aux;
    while (primero != nullptr) {
        aux = primero;
        primero = primero->getSiguiente();
        delete aux;
    }
}

void ListaPropietarios::agregarPropietario(const Propietario &propietario) {
    Nodo<Propietario>* nuevoNodo = new Nodo<Propietario>(propietario);

    if (primero == nullptr) {
        primero = nuevoNodo;
        ultimo = nuevoNodo;
    } else {
        ultimo->setSiguiente(nuevoNodo);
        nuevoNodo->setAnterior(ultimo);
        ultimo = nuevoNodo;
    }
}

Propietario* ListaPropietarios::buscarPropietarioPorCedula(const string &cedula) {
    Nodo<Propietario>* aux = primero;
    while (aux != nullptr) {
        if (aux->getDato().getCedula() == cedula) {
            return &aux->getDato();
        }
        aux = aux->getSiguiente();
    }
    return nullptr;
}

void ListaPropietarios::mostrarPropietarios() const {
    Nodo<Propietario>* aux = primero;
    if (aux == nullptr) {
        cout << "La lista de propietarios está vacía." << endl;
        return;
    }
    while (aux != nullptr) {
        cout << aux->getDato().toString() << endl;
        aux = aux->getSiguiente();
    }
}

bool ListaPropietarios::estaVacia() const {
    return primero == nullptr;
}

void ListaPropietarios::cargarArchivo(const string &archivo) {
    ifstream file(archivo);
    if (file.is_open()) {
        string linea;
        while (getline(file, linea)) {
            stringstream ss(linea);
            string nombre, apellido, cedula, correo, placa;
            getline(ss, nombre, ',');
            getline(ss, apellido, ',');
            getline(ss, cedula, ',');
            getline(ss, correo, ',');
            Propietario propietario(nombre, apellido, cedula, correo);
            while (getline(ss, placa, ',')) {
                propietario.agregarPlaca(ListaDoble<Propietario>::descifrarTexto(placa));
            }
            agregarPropietario(propietario);
        }
        file.close();
    } else {
        cerr << "Error al abrir el archivo para cargar propietarios." << endl;
    }
}

void ListaPropietarios::guardarArchivo(const string &archivo) const {
    ofstream file(archivo);
    if (file.is_open()) {
        Nodo<Propietario>* aux = primero;
        while (aux != nullptr) {
            const Propietario& propietario = aux->getDato();
            file << propietario.getNombre() << ","
                 << propietario.getApellido() << ","
                 << propietario.getCedula() << ","
                 << propietario.getCorreo();
            for (const auto &placa : propietario.getPlacas()) {
                file << "," << ListaDoble<Propietario>::cifrarTexto(placa);
            }
            file << endl;
            aux = aux->getSiguiente();
        }
        file.close();
    } else {
        cerr << "Error al abrir el archivo para guardar propietarios." << endl;
    }
}

File: ESTRUCTURA DE DATOS/DEBERES/SEGUNDO_PARCIAL/DEBER2/PROYECTO_NRC1978/LISTA_SIMPLE_PYTHON/PYTHON/lista_propietarios.py
from nodo import Nodo
from propietario import Propietario

class ListaPropietarios:
    def __init__(self):
        self.__primero = None
        self.__ultimo = None

    def __del__(self):
        aux = self.__primero
        while aux is not None:
            siguiente = aux.get_siguiente()
            del aux
            aux = siguiente

    def agregar_propietario(self, propietario):
        nuevo_nodo = Nodo(propietario)
        if self.__primero is None:
            self.__primero = nuevo_nodo
            self.__ultimo = nuevo_nodo
        else:
            self.__ultimo.set_siguiente(nuevo_nodo)
            self.__ultimo = nuevo_nodo

    def buscar_propietario_por_cedula(self, cedula):
        aux = self.__primero
        while aux is not None:
            if aux.get_dato().get_cedula() == cedula:
                return aux.get_dato()
            aux = aux.get_siguiente()
        return None

    def mostrar_propietarios(self):
        aux = self.__primero
        if aux is None:
            print("La lista de propietarios está vacía.")
            return
        while aux is not None:
            print(aux.get_dato().to_string())
            aux = aux.get_siguiente()

    def esta_vacia(self):
        return self.__primero is None

    def cargar_archivo(self, archivo):
        try:
            with open(archivo, 'r') as file:
                for linea in file:
                    partes = linea.strip().split(',')
                    if len(partes) < 4:
                        continue
                    nombre, apellido, cedula, correo = partes[:4]
                    propietario = Propietario(nombre, apellido, cedula, correo)
                    for placa in partes[4:]:
                        propietario.agregar_placa(placa)
                    self.agregar_propietario(propietario)
        except FileNotFoundError:
            print("Error: No se pudo abrir el archivo para cargar propietarios.")
        except Exception as e:
            print(f"Error inesperado al cargar el archivo: {e}")

    def guardar_archivo(self, archivo):
        try:
            with open(archivo, 'w') as file:
                aux = self.__primero
                while aux is not None:
                    propietario = aux.get_dato()
                    line = f"{propietario.get_nombre()},{propietario.get_apellido()},{propietario.get_cedula()},{propietario.get_correo()}"
                    for placa in propietario.get_placas():
                        line += f",{placa}"
                    file.write(line + '\n')
                    aux = aux.get_siguiente()
        except Exception as e:
            print(f"Error al guardar el archivo: {e}")
